### PR TITLE
Change network message format to bincode instead of json

### PR DESCRIPTION
**Pull request by Emiluren**

### CODE REVIEW

**Comment by Emiluren** [comment]
Fixes #23

**Review by malcx95 — APPROVED** [review]
Looks good, seems like you just swapped serde for bincode

File: src/messages.rs [hunk 1]
```diff
@@ -46,27 +46,29 @@ macro_rules! impl_message_reader {
...
+                ]) as usize;
 
-                let as_str = String::from_utf8_lossy(&msg_bytes);
+                // We will not read a message until a complete message has been
+                // received
+                if self.byte_queue.len() < 2 + length {
+                    return None;
+                }
+
+                self.byte_queue.pop_front().unwrap();
```
Comment (malcx95) [inline src/messages.rs:65]: Why is this done?

Comment (Emiluren) [reply src/messages.rs:65]: The length of the rest of the message is stored in the first two bytes

File: src/client.rs [hunk 2]
```diff
@@ -50,11 +50,11 @@ impl KeyStates {
...
 
 fn send_client_message(msg: &ClientMessage, stream: &mut TcpStream) {
-    let data = serde_json::to_string(msg)
-        .expect("Failed to encode message");
-    stream.write(data.as_bytes())
-        .expect("Failed to send message to server");
-    stream.write(&[0])
+    let data = bincode::serialize(msg).expect("Failed to encode message");
+    let length = data.len() as u16;
+    stream.write(&length.to_be_bytes())
```
Comment (TheZoq2) [inline src/client.rs:55]: Might be better to send the length and the data together as one message in case one succeeds and the other one fails

**Review by HannesTuhkala — APPROVED** [review]

File: src/messages.rs (hunk 1 above)
Comment (TheZoq2) [reply src/messages.rs:65]: But why are the bytes popped and not looked at?

File: src/client.rs (hunk 2 above)
Comment (Emiluren) [reply src/client.rs:55]: Tcp doesn't have a concept of messages. It's stream oriented and bufffered. Besides we already split single "messages" on the server side if we don't send everything at once.

File: src/messages.rs (hunk 1 above)
Comment (Emiluren) [reply src/messages.rs:65]: They are looked at. When we read the length out. But we don't want to pop them until we are sure that our entire message has arrived so we can read it all at once.

**Pull request merged**